### PR TITLE
drivers: counter: add driver for NXP QTMR counters using MCUXpresso SDK

### PR DESCRIPTION
**Pull request by kt-peters**
This includes the driver and a modification of the counter_basic_api test to test it. The driver requires a change in the hal_nxp repository to work. It has been tested on a mimxrt1060_evk board .

Signed-off-by: Jan Peters <peters@kt-elektronik.de>

### CODE REVIEW

**Comment by kt-peters** [comment]
The driver requires https://github.com/zephyrproject-rtos/hal_nxp/pull/164 to work.

**Comment by carlescufi** [comment]
> The driver requires [zephyrproject-rtos/hal_nxp#164](https://github.com/zephyrproject-rtos/hal_nxp/pull/164) to work.

Can you then point to this PR in the `west.yml` entry? See:
https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules

File: dts/arm/nxp/nxp_rt.dtsi [hunk 1]
```diff
@@ -120,6 +120,134 @@
 			label = "GPT2";
 		};
 
+		tmr1: qtmr@401dc000 {
```
Comment (danieldegrasse) [inline dts/arm/nxp/nxp_rt.dtsi:123]: I'd prefer to use labels like `qmtr1` here (as opposed to `tmr1), to follow the pattern used by the remainder of the nodes in the DTSI file.

File: dts/bindings/counter/nxp,imx-tmr.yaml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      - 2
+      - 3
+
+  mode:
+    type: int
+    required: true
+    description: counting mode of the timer, see qtmr_counting_mode_t enumerator type
+      of the MCUXpresso SDK
+    enum:
+      - 0
```
Comment (danieldegrasse) [inline dts/bindings/counter/nxp,imx-tmr.yaml:30]: Can you use a string type for this property? `DT_ENUM_IDX()` can then be used to get the underlying enum value.

File: dts/bindings/counter/nxp,imx-tmr.yaml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      - 6
+      - 7
+
+  primary_source:
+    type: int
+    required: true
+    description: Primary source of the timer, see qtmr_primary_count_source_t enumerator type
+      of the MCUXpresso SDK
+    enum:
+      - 0
```
Comment (danieldegrasse) [inline dts/bindings/counter/nxp,imx-tmr.yaml:45]: Same comment about string type with `DT_ENUM_IDX`

File: dts/bindings/counter/nxp,imx-tmr.yaml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      - 14
+      - 15
+
+  secondary_source:
+    type: int
+    required: false
+    description: Secondary source of the timer, see qtmr_input_source_t enumerator type
+      of the MCUXpresso SDK
+    enum:
+      - 0
```
Comment (danieldegrasse) [inline dts/bindings/counter/nxp,imx-tmr.yaml:68]: Ditto: use string type and `DT_ENUM_IDX`

File: tests/drivers/counter/counter_basic_api/testcase.yaml [hunk 5]
```diff
@@ -11,4 +11,4 @@ tests:
     platform_allow: nrf52840dk_nrf52840
     timeout: 400
     extra_args: >
-      OVERLAY_CONFIG="zli.conf;boards/nrf52840dk_nrf52840_zli.conf"
+      OVERLAY_CONFIG="zli.conf;boards/nrf52840dk_nrf52840_zli.conf;boards/mimxrt1060_evk.conf"
```
[the remark this answers is not in the document]
Comment (kt-peters) [reply tests/drivers/counter/counter_basic_api/testcase.yaml]: okay, I removed it.

File: dts/arm/nxp/nxp_rt.dtsi (hunk 1 above)
Comment (kt-peters) [reply dts/arm/nxp/nxp_rt.dtsi]: Changed.

File: dts/bindings/counter/nxp,imx-tmr.yaml (hunk 2 above)
Comment (kt-peters) [reply dts/bindings/counter/nxp,imx-tmr.yaml]: Thanks, I didn't know that macro yet. It's very useful. Changed accordingly.

**Review by danieldegrasse — DISMISSED** [review]
Thanks for the contribution!

**Comment by dleach02** [comment]
@kt-peters the NXP HAL PR has been merged. Please update your west.yaml so we can merge this PR

File: drivers/counter/counter_mcux_qtmr.c [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+	/* the interrupt can be triggered by any of the four channels of the QTMR. Check status
+	 * of all channels and trigger the ISR for the channel(s) that has/have triggered the
+	 * interrupt.
+	 */
+	for (qtmr_channel_selection_t ch = kQTMR_Channel_0; ch <= kQTMR_Channel_3 ; ch++) {
+		if (timers[ch] != NULL) {
+			const struct mcux_qtmr_config *config = timers[ch]->config;
+			struct mcux_qtmr_data *data = timers[ch]->data;
+
+			uint32_t channel_status = QTMR_GetStatus(config->base, ch);
```
Comment (mmahadevan108) [inline drivers/counter/counter_mcux_qtmr.c:92]: Seems unnecessary to have this in the for loop as you need to get the status reg only once.

Comment (kt-peters) [reply drivers/counter/counter_mcux_qtmr.c:92]: actually, I have to get the status once for every channel (note that the loop variable `ch` is an argument to this function call).

File: tests/drivers/counter/counter_basic_api/boards/mimxrt1060_evk.conf [hunk 7]
```diff
@@ -0,0 +1 @@
+CONFIG_COUNTER_MCUX_QTMR=y
```
Comment (mmahadevan108) [inline tests/drivers/counter/counter_basic_api/boards/mimxrt1060_evk.conf:1]: Please pull all test changes to a separate commit

File: tests/drivers/counter/counter_basic_api/src/test_counter.c [hunk 8]
```diff
@@ -800,6 +803,9 @@ static void test_short_relative_alarm_instance(const char *dev_name)
 		.user_data = NULL
 	};
 
+	/* for timers with very short ticks, counter_ticks_to_us() returns 0 */
```
[the remark this answers is not in the document]
Comment (kt-peters) [reply tests/drivers/counter/counter_basic_api/src/test_counter.c:805]: In this specific case, the call is `counter_ticks_to_us(dev, 1)` - the clock supplying the counter on the mimxrt1060_evk board runs at 150 MHz and is configured in the overlay to use a divider of 128 (the highest divider available for the QTMR), so one tick is 128/(150 MHz) = 0.85 us - which is rounded to 0 in the integer calculation. Calling `counter_ticks_to_us()` for more than 1 ticks results at least 1 us. The only other case where `counter_ticks_to_us(dev, 1)` is called is `test_late_alarm_error`, which is skipped for the QTMR, as the driver does not support setting a guard period which appears to be a prerequisite of the test.

File: drivers/counter/counter_mcux_qtmr.c (hunk 6 above)
Comment (mmahadevan108) [reply drivers/counter/counter_mcux_qtmr.c:92]: Thank you. I missed that there is a status register per channel

File: west.yml [hunk 9]
```diff
@@ -101,7 +101,7 @@ manifest:
       groups:
         - hal
     - name: hal_nxp
-      revision: 9ac27f746517a483c01df2721957c822b5bda954
+      revision: d287347745db7ccc85d267fddf3ce4f89a98ec1f
```
[the remark this answers is not in the document]
Comment (kt-peters) [reply west.yml:104]: Done.

File: drivers/counter/counter_mcux_qtmr.c [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+	}
+
+	return 0;
+}
+
+static uint32_t mcux_qtmr_get_top_value(const struct device *dev)
+{
+	const struct mcux_qtmr_config *config = dev->config;
+
+	return config->info.max_top_value;
```
Comment (mmahadevan108) [inline drivers/counter/counter_mcux_qtmr.c:244]: If `set_top_value` is set, then this should return that as the `top_value`

Comment (kt-peters) [reply drivers/counter/counter_mcux_qtmr.c:244]: The QTMR does not support setting a top value other than the one stored in `config->info.max_top_value` (which is 0xffff).

The implementation of `set_top_value` for this driver returns `-ENOTSUP` for any other value. As far as I understand, that's the default behaviour in this case.

Comment (mmahadevan108) [reply drivers/counter/counter_mcux_qtmr.c:244]: Given there are 4 channels, could we reserve one of the channels for `set_top_value` implementation and take out one from the device tree file.

**Comment by mmahadevan108** [comment]
thank you for the contribution.

**Review by mmahadevan108 — DISMISSED** [review]

**Review by danieldegrasse — COMMENTED** [review]
@kt-peters, can you move the commit updating the NXP HAL to be the first commit in the series, and change the commit message to something like the following:

"manifest: hal_nxp: update hal revision to pull in mcux_qtmr driver"

**Review by mmahadevan108 — APPROVED** [review]

**Review by danieldegrasse — APPROVED** [review]

**Comment by dleach02** [comment]
@kt-peters, thank you for your contribution
